File: backend/app/models/transaction.rb
class Transaction < ApplicationRecord
  STATUSES = %w[pending completed rejected].freeze
  KINDS = %w[exchange deposit recharge transfer].freeze

  belongs_to :user

  validates :source_currency, :target_currency, :source_amount,
            :target_amount, :exchange_rate, :status, :kind, presence: true
  validates :source_amount, :target_amount, :exchange_rate,
            numericality: { greater_than_or_equal_to: 0 }
  validates :status, inclusion: { in: STATUSES }
  validates :kind, inclusion: { in: KINDS }
  validates :source_currency, :target_currency, inclusion: { in: Wallet::CURRENCIES }
end

File: frontend/src/types/transaction.ts
import type { Currency } from "./wallet";

export type TransactionStatus = "pending" | "completed" | "rejected";
export type TransactionKind = "exchange" | "deposit" | "recharge" | "transfer";

export interface Transaction {
  id: number;
  kind: TransactionKind;
  source_currency: Currency;
  target_currency: Currency;
  source_amount: string;
  target_amount: string;
  exchange_rate: string;
  status: TransactionStatus;
  rejection_reason: string | null;
  created_at: string;
}
